common/djangoapps/course_modes/views.py

```diff
@@ -101,10 +101,18 @@ def get(self, request, course_id, error=None):
                     redirect_url = ecommerce_service.checkout_page_url(professional_mode.bulk_sku)
             return redirect(redirect_url)
 
+        # If the learner has arrived at this screen via the traditional enrollment workflow, then they should
+        # already be enrolled in an audit mode for the course, assuming one has been configured.  However,
+        # alternative enrollment workflows have been introduced into the system, such as third-party discovery.
+        # These workflows result in learners arriving directly at this screen, and they will not necessarily
+        # be pre-enrolled in the audit mode.  In that case we'll need to enroll them here before going further
+        # and then reload the state of the "enrollment_mode" and "is_active" parameters.
+        if not is_active and (not enrollment_mode or enrollment_mode == CourseMode.AUDIT):
```

Comment:
I'm not sure this is the right logical check. I ran into trouble testing this because the user I was using had previously been enrolled in a premium course mode. The enrollment wasn't active, meeting the first condition, but it was a premium mode, failing the second condition. As a result, I wasn't automatically enrolled in an audit mode when visiting the page.

In the case where the premium mode still exists, I'm taken to the view, and I can choose to enroll in an audit mode, which works. However, if the premium course mode has been removed, I'm redirected to the Dashboard, without any enrollment being set up for me.

I'm actually not sure if we need something beyond the `is_active` check. Can you clarify the need for the `enrollment_mode` checks? Obviously we wouldn't want to overwrite an _active_ premium enrollment, but I'm not sure if we want to avoid overwriting an _inactive_ premium mode enrollment.